larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+            maxFitGapLengthCluster           /= straightLineLengthLargeCluster;
+            rmsSlidingLinearFitCluster       /= straightLineLengthLargeCluster;
+        }
+
+        diffWithStraightLineMean   += diffWithStraightLineMeanCluster;
+        diffWithStraightLineSigma  += diffWithStraightLineSigmaCluster;
+        dTdLWidth                  += dTdLWidthCluster;
+        maxFitGapLength            += maxFitGapLengthCluster;
+        rmsSlidingLinearFit        += rmsSlidingLinearFitCluster;
+    }
+
+    const int nClusters(clusterList.size());
+    diffWithStraightLineMean   /= nClusters;
+    diffWithStraightLineSigma  /= nClusters;
+    dTdLWidth                  /= nClusters;
+    maxFitGapLength            /= nClusters;
+    rmsSlidingLinearFit        /= nClusters;
+
+    if(m_addStraightLineLength)
+    {
+        float lenghtSquare(LArPfoHelper::GetThreeDLengthSquared(pInputPfo));
+        featureVector.push_back(std::sqrt(lenghtSquare));
+    }
+    if (m_addDiffWithStraightLineMean)
+        featureVector.push_back(diffWithStraightLineMean);
+
+    if (m_addDiffWithStraightLineSigma)
+        featureVector.push_back(diffWithStraightLineSigma);
+
+    if (m_addDTDLWidth)
+        featureVector.push_back(dTdLWidth);
+
+    if (m_addMaxFitGapLength)
+        featureVector.push_back(maxFitGapLength);
+
+    if (m_addRMSLinearFit)
+        featureVector.push_back(rmsSlidingLinearFit);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeDLinearFitFeatureTool::CalculateVariablesSlidingLinearFit(const pandora::Cluster *const pCluster, float &straightLineLengthLarge,
+    float &diffWithStraightLineMean, float &diffWithStraightLineSigma, float &dTdLWidth, float &maxFitGapLength, float &rmsSlidingLinearFit) const
+{
+    try
+    {
+        const TwoDSlidingFitResult slidingFitResult(pCluster, m_slidingLinearFitWindow, LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+        const TwoDSlidingFitResult slidingFitResultLarge(pCluster, m_slidingLinearFitWindowLarge, LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+
+        if (slidingFitResult.GetLayerFitResultMap().empty())
+            throw StatusCodeException(STATUS_CODE_NOT_INITIALIZED);
+
+        straightLineLengthLarge = (slidingFitResultLarge.GetGlobalMaxLayerPosition() - slidingFitResultLarge.GetGlobalMinLayerPosition()).GetMagnitude();
+        rmsSlidingLinearFit = 0.f;
+
+        FloatVector diffWithStraightLineVector;
+        const HitType hitType(LArClusterHelper::GetClusterHitType(pCluster));
+        CartesianVector previousFitPosition(slidingFitResult.GetGlobalMinLayerPosition());
+        float dTdLMin(+std::numeric_limits<float>::max()), dTdLMax(-std::numeric_limits<float>::max());
+
+        for (const auto &mapEntry : slidingFitResult.GetLayerFitResultMap())
+        {
+            rmsSlidingLinearFit += mapEntry.second.GetRms();
+
```

Comment:
Might want to cache mapEntry.second as a const reference to a LayerFitResult here, just to improve readability.

Reply:
ok